packages/wasm/crate/src/tx.rs

```diff
@@ -273,7 +276,89 @@ pub async fn transaction_info_inner(
                         .insert(nullifier, spendable_note_record.note.clone());
                 }
             }
+            Action::Swap(swap) => {
```

Comment:
There may be a neater, less nest-y way of writing all this, but that is outside of my Rust knowledge 😆

Reply:
What do you think of this?

```rust
Action::Swap(swap) => {
                let commitment = swap.body.payload.commitment;

                let swap_position_option = storage
                    .get_swap_by_commitment(commitment.into())
                    .await?
                    .map(|swap_record| Position::from(swap_record.position));

                if let Some(swap_position) = swap_position_option {
                    add_swap_info_to_perspective(&storage, &fvk, &mut txp, &commitment, swap_position).await?;
                }
            }
```

and then an external fn
```rust

async fn add_swap_info_to_perspective(storage: &IndexedDBStorage, fvk: &FullViewingKey, txp: &mut penumbra_transaction::TransactionPerspective, commitment: &StateCommitment, swap_position: Position) -> Result<(), WasmError> {
    let derived_nullifier_from_swap =
        Nullifier::derive(fvk.nullifier_key(), swap_position, commitment);

    let transaction_infos = storage.get_transaction_infos().await?;

    for transaction_info in transaction_infos {
        transaction_info
            .transaction
            .and_then(|transaction| transaction.body)
            .iter()
            .for_each(|body| {
                for action in body.actions.iter() {
                    let transaction_id = action.action.as_ref()
                        .and_then(|action| {
                            match action {
                                v1::action::Action::SwapClaim(swap_claim) => swap_claim.body.as_ref(),
                                _ => None,
                            }
                        })
                        .and_then(|body| body.nullifier.as_ref())
                        .filter(|&nullifier| nullifier == &derived_nullifier_from_swap.to_proto())
                        .and(transaction_info.id.as_ref())
                        .and_then(|id| TransactionId::try_from(id.clone()).ok());

                    if let Some(transaction_id) = transaction_id {
                        txp.nullification_transaction_ids_by_commitment.insert(*commitment, transaction_id);
                        break;
                    }
                }
            });
    }
    Ok(())
}

```

Reply:
Much better — thanks!